Redesign Javalin tests

NicMcPhee opened the pull request:
This is basically a spike to see what things would look like if we followed the pattern in the Javalin testing examples for more "purely" unit tests. It's sitting on top of #22 because it doesn't really make a lot of sense for it to be on `main` without all the Javalin 4 changes.

So far all this does is add a new `getUsersByAge()` test that follows that pattern, but that's enough to start a conversation.

I brought in Mockito to mock the `Context` class and then spec'ed what the mock context would return when `queryParamMap()` and `queryParamAsClass()` were called.

These `when().thenReturn()` are more complicated that I'd like in part because by default Mockito won't allow you to mock chained calls. You *can* do it, but they strongly discourage it (https://www.javadoc.io/doc/org.mockito/mockito-core/2.13.0/org/mockito/Mockito.html#RETURNS_DEEP_STUBS), so I didn't want to immediately go there.

So in a lot of ways I'd rather have something like:

```java
   when(ctx.queryParamMap().contains("age")).thenReturn(true)
```

To do that in a way that the Mockito people would like, I think we'd need to make a subclass of Context that has direct methods for things like seeing if there's an "age" key. Seems a bit heavy to me at the moment, but I could be talked into trying it.

Rather than capturing the argument to `ctx.json()`, I used the Mockito `ArgumentMatcher`, which is what the Mockito folks recommend that you use instead of argument capturing. That takes a function that checks the argument in question. In our case this means looping through the list and asserting that everyone has age 37.

NicMcPhee committed:
Redesign unit test for get all users by age

This adds a new `getUsersByAge()` test that follows the pattern in the Javalin testing examples for more "purely" unit tests.

I brought in Mockito to mock the `Context` class and then spec'ed what the mock context would return when `queryParamMap()` and `queryParamAsClass()` were called.

These `when().thenReturn()` are more complicated that I'd like in part because by default Mockito won't allow you to mock chained calls. You *can* do it, but they strongly discourage it (https://www.javadoc.io/doc/org.mockito/mockito-core/2.13.0/org/mockito/Mockito.html#RETURNS_DEEP_STUBS), so I didn't want to immediately go there.

So in a lot of ways I'd rather have something like:

   when(ctx.queryParamMap().contains("age")).thenReturn(true)

To do that in a way that the Mockito people would like, I think we'd need to make a subclass of Context that has direct methods for things like seeing if there's an "age" key. Seems a bit heavy to me at the moment, but I could be talked into trying it.

Rather than capturing the argument to `ctx.json()`, I used the Mockito `ArgumentMatcher`, which is what the Mockito folks recommend that you use instead of argument capturing. That takes a function that checks the argument in question. In our case this means looping through the list and asserting that everyone has age 37.

NicMcPhee commented on server/src/main/java/umm3601/user/UserController.java:
```diff
@@ -29,9 +29,9 @@
-  private static final String AGE_KEY = "age";
-  private static final String COMPANY_KEY = "company";
-  private static final String ROLE_KEY = "role";
+  public static final String AGE_KEY = "age";
```
I made these variable public so the test code could access them.

NicMcPhee commented on server/src/test/java/umm3601/user/UserControllerSpec.java:
```diff
@@ -152,21 +196,53 @@ public void GetUsersByAge() throws IOException {
   }
 
+  @Test
+  public void getUsersByAge() throws JsonMappingException, JsonProcessingException {
```
This is the new method for discussion.

NicMcPhee committed:
Merge branch 'main' into redesign-javalin-tests

NicMcPhee committed:
Merge branch 'main' into redesign-javalin-tests

NicMcPhee committed:
Merge branch 'main' into redesign-javalin-tests

kklamberty commented on server/src/test/java/umm3601/user/UserControllerSpec.java:
```diff
@@ -203,6 +211,38 @@ public void GetUsersByAge() throws IOException {
+  @Test
+  public void getUsersByAge() throws JsonMappingException, JsonProcessingException {
+    Context ctx = mock(Context.class);
+    // When the controller calls `ctx.queryParamMap`, return the expected map for an
+    // "?age=37" query.
+    when(ctx.queryParamMap()).thenReturn(Map.of(UserController.AGE_KEY, List.of("37")));
+    // When the controller calls `ctx.queryParamAsClass() to get the value associated with
+    // the "age" key, return an appropriate Validator. TBH, I never did figure out what the
+    // third argument to the Validator constructor was for, but `null` seems OK. I'm also not sure
+    // what the first argument is; it appears that you can set it to anything that isn't
+    // null and it's happy.
+    Validator<Integer> validator = new Validator<Integer>("age", 37, null);
+    when(ctx.queryParamAsClass(UserController.AGE_KEY, Integer.class)).thenReturn(validator);
```
In another commit #https://github.com/UMM-CSci-3601/3601-iteration-template/commit/8a7328baf9fdbf4dfa3f6206e2aeae76932994ea we have something similar that creates a validator in a different way:
```typescript
when(ctx.queryParamAsClass("age", Integer.class))
      .thenReturn(Validator.create(Integer.class, "37", "age"));
```

NicMcPhee replied:
These are pretty much the same. I like the simpler construction of the `Validator` in the second version, but I probably like giving it a name like we do in the first version. I also definitely prefer having `UserController.AGE_KEY` instead of magic strings everywhere.

kklamberty reviewed (commented):
I wonder if we should look at this alongside the changes we made in #https://github.com/UMM-CSci-3601/3601-iteration-template/commit/8a7328baf9fdbf4dfa3f6206e2aeae76932994ea (as an example) to also verify that the users each have age 37. I'd like to discuss this a bit more.

NicMcPhee commented on server/src/test/java/umm3601/user/UserControllerSpec.java:
```diff
@@ -203,6 +211,38 @@ public void GetUsersByAge() throws IOException {
+    // Verify that `getUsers` called `ctx.status(200)` at some point.
+    verify(ctx).status(200);
+
+    // Verify that `ctx.json()` is called with a `List` of `User`s.
+    // Each of those `User`s should have age 37.
+    verify(ctx).json(argThat(new ArgumentMatcher<List<User>>() {
+      public boolean matches(List<User> users) {
+        for (User user : users) {
+          assertEquals(37, user.age);
+        }
+        return true;
+      }
+    }));
```
I think this is a test of the controller. We're saying that if the request has the age being 37, then all the results should be 37. That does indirectly test the database, but it also requires that we make the right query to the database, etc., which is what this really tests.

We could actually mock the DB as well if we wanted to take it completely out of the unit testings loop, but that would be a pretty substantial change.

kklamberty replied:
This PR has been updated to include this new test along with the other new tests. I kept this one even though it's a bit different at least in part because of the use of the argument matcher.

kklamberty committed:
Merge branch 'main' into redesign-javalin-tests

kklamberty committed:
Incorporate a way of testing by age that was explored by Nic

Nic made a branch and PR exploring a different way to do the Javalin tests. Some of this was addressed in a major overhaul of how the Javalin testing was done, but there are some things Nic tried in this branch that seemed like they might be helpful to retain.

This commit also involves removing some magic numbers by replacing them with constants. Those constants need to be available to the package for them to be used in these tests.

kklamberty committed:
Remove unused imports

kklamberty committed:
Fix minor differences between Nic's test and other tests

One thing that needed to change was that we didn't want to have a local mock context since we have one for the whole class. The other thing was that we are using HttpStatus.OK rather than just the number 200.

kklamberty commented on server/.vscode/settings.json:
```diff
@@ -0,0 +1,15 @@
+{
+    "cSpell.words": [
+        "Bson",
+        "invalidemail",
+        "invalidrole",
+        "Javalin",
+        "notanumber",
+        "OHMNET",
+        "sams",
+        "Servlet",
+        "sortorder",
+        "todos"
+    ],
+    "java.configuration.updateBuildConfiguration": "interactive"
+}
```
Do we want to generate a .vscode folder and put these words in the spell checker?

NicMcPhee replied:
Isn't there already a `.vscode` folder with the `cSpell.words` entry?

NicMcPhee replied:
Maybe there isn't. If not, this will make that if/when we merge it in, and we should be good.

kklamberty commented on server/build.gradle:
```diff
@@ -58,6 +58,9 @@ dependencies {
   testImplementation 'org.mockito:mockito-core:5.1.1'
 
   testImplementation 'com.mockrunner:mockrunner-servlet:2.0.6'
+
+  // The Mockito Java mocking library
+  testImplementation "org.mockito:mockito-core:3.+"
```
This would have been new when it was first a branch... but it's not going to write over the newer version here, is it?

NicMcPhee replied:
It won't _overwrite_ it, but it will conflict with the 5.1.1 definition on line 58. We definitely don't want these lines.

kklamberty commented on server/src/main/java/umm3601/user/UserController.java:
```diff
@@ -33,11 +33,11 @@
-  private static final String COMPANY_KEY = "company";
-  private static final String ROLE_KEY = "role";
-  private static final int REASONABLE_AGE_LIMIT = 150;
+  static final String AGE_KEY = "age";
+  static final String COMPANY_KEY = "company";
+  static final String ROLE_KEY = "role";
```
Nic originally made these public, but that's a little more open than is needed. Package level access is enough.

NicMcPhee replied:
People often make "constants" public, and I think I just did that out of habit.

kklamberty commented on server/src/main/java/umm3601/user/UserController.java:
```diff
@@ -33,11 +33,11 @@
+  static final String COMPANY_KEY = "company";
+  static final String ROLE_KEY = "role";
 
+  private static final int REASONABLE_AGE_LIMIT = 150;
```
This was new since the branch was created

kklamberty commented on server/src/test/java/umm3601/user/UserControllerSpec.java:
```diff
@@ -198,10 +204,10 @@ public void canGetAllUsers() throws IOException {
     // Add a query param map to the context that maps "age" to "37".
     Map<String, List<String>> queryParams = new HashMap<>();
-    queryParams.put("age", Arrays.asList(new String[] {"37"}));
+    queryParams.put(UserController.AGE_KEY, Arrays.asList(new String[] {"37"}));
```
This and many other places that follow were changing from magic numbers to specific values.

NicMcPhee replied:
We probably want to give names to the string `"37"`, or make that by calling `.toString()` on the value of `AGE_KEY`. Otherwise we've only kinda done half the job.

kklamberty reviewed (approved):
@NicMcPhee, you may want to review this to see if you want this to be merged.

NicMcPhee committed:
Merge branch 'main' into redesign-javalin-tests

NicMcPhee commented on server/build.gradle:
```diff
@@ -58,6 +58,9 @@ dependencies {
   testImplementation 'org.mockito:mockito-core:5.1.1'
 
   testImplementation 'com.mockrunner:mockrunner-servlet:2.0.6'
+
+  // The Mockito Java mocking library
+  testImplementation "org.mockito:mockito-core:3.+"
```
```suggestion

```

NicMcPhee commented on server/src/test/java/umm3601/user/UserControllerSpec.java:
```diff
@@ -198,10 +204,10 @@ public void canGetAllUsers() throws IOException {
   public void canGetUsersWithAge37() throws IOException {
```
I think my intent had been to get rid of this _in favor of_ the (then new) `getUsersByAge()` method. Is there a reason to keep this? If we do, students will just copy and paste it all over Lab 4 and then their projects.

NicMcPhee reviewed (commented):
I don't think this is really ready for prime time, yet. There are several different approaches to the testing in the various `getX()` tests, and I think that'll just confuse the students. I think my goal had been to take the idea in `getUsersByAge()` and, _if_ people thought it was good, then copy that out across the other tests like `getUsersByRole()`. As it is, we have about 1/4 of the job done.
